File: slurmemu/src/timer.rs


pub struct Timer {

	pub en: bool,
	pub count: u32,
	pub count_out: u16,
	pub match_val: u16,
}

impl Timer {

    pub fn new() -> Self {
        Timer { 
		en: false,
		count: 0,
		count_out: 0,
		match_val: 0
	}
    }

    pub fn port_op(& mut self, _port : u16, val: u16, write : bool) -> u16
    {
	let mut ret : u16 = 0;

        if write {
        	match _port & 0xf
		{
			0 =>	self.en = val & 1 == 1,
			1 =>	self.match_val = val,
			_ =>  (),
		}
	}

        return self.count_out;
    }

    pub fn step(& mut self) -> bool
    {
	if self.en
	{
		self.count = (self.count + 1) & 0x007fffff;
	}
	self.count_out = (self.count >> 7) as u16;
	return self.count_out == self.match_val;
    }

}

File: slurmemu-ng/src/Timer.cpp
/* vim: set et ts=4 sw=4: */

/*
	slurmemu-ng : Next-Generation SlURM16 Emulator

Timer.cpp: Timer emulator

License: MIT License

Copyright 2023 J.R.Sharp

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

*/

#include "Timer.h"

#include <cstdio>

Timer::Timer()  :
    m_en(false),
    m_match_val(0x1234)
{

}

Timer::~Timer()
{

}

std::uint16_t Timer::port_op(std::uint16_t port, bool write, std::uint16_t wr_val)
{
    if (write)
    {
        switch (port & 0xf)
        {
            case 0:
                m_en = !!(wr_val & 1);
                break;
            case 1:
                m_match_val = wr_val;
                break;
        }
    }
    
    return m_count_out;
} 
   
bool Timer::step(std::uint16_t* mem)
{
    if (m_en)
        m_count = (m_count + 1) & 0x007fffff;

    m_count_out = (m_count >> 7);

    bool intrpt = (m_en && (m_count_out == m_match_val));

    return intrpt;
}

